### PR TITLE
DEM-1207 Update Type Controller for new Data Model

### PR DESCRIPTION
**Pull request by samsep**
ping @appirio-tech/demandside

### CODE REVIEW

File: src/scripts/controllers/type.controller.coffee [hunk 1]
```diff
@@ -1,89 +1,163 @@
...
+    updates = getUpdates()
+    if workValid
+      SubmitWorkService.save(updates).then ->
+        vm.showSuccessModal = true
+
+  typeValid = ->
+    updates = getUpdates()
+    isValid = true
+    for type, value of updates
+      if !value.length
```
Comment (vietnogi) [inline src/scripts/controllers/type.controller.coffee]: `isValid = false unless value.length`

File: src/scripts/controllers/type.controller.coffee [hunk 2]
```diff
@@ -1,89 +1,163 @@
...
+
+    updates = getUpdates()
 
   activate = ->
-    if vm.workId
-      params =
-        id: vm.workId
-
-      resource = SubmitWorkAPIService.get params
+    destroyWorkListener = $rootScope.$on "SubmitWorkService.work:changed", ->
```
Comment (vietnogi) [inline src/scripts/controllers/type.controller.coffee]: lets avoid using `$rootScope.$on`, an alternative option is to use `$scope.$watch`

File: src/scripts/services/submit-work.service.coffee [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      SubmitWorkAPIService.get(params).$promise
+
+    Optimist.fetchOne {
+      model: submitWorkService.work
+      apiCall: apiCall
+      updateCallback: emitUpdates
+    }
+
+  submitWorkService.save = (updates) ->
+    console.log updates
```
Comment (vietnogi) [inline src/scripts/services/submit-work.service.coffee]: remove

File: src/scripts/services/submit-work.service.coffee [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+mockWork = {}
+srv = ($rootScope, Optimist, SubmitWorkAPIService) ->
+  # Used for caching
+  currentWorkId = null
+
+  submitWorkService =
+    work: {}
+
+  emitUpdates = ->
+    $rootScope.$emit 'SubmitWorkService.work:changed'
```
Comment (vietnogi) [inline src/scripts/services/submit-work.service.coffee]: lets avoid `$rootScope.$emit 'SubmitWorkService.work:changed'`

File: src/scripts/services/submit-work.service.coffee [hunk 5]
```diff
@@ -0,0 +1,49 @@
+'use strict'
+mockWork = {}
+srv = ($rootScope, Optimist, SubmitWorkAPIService) ->
+  # Used for caching
+  currentWorkId = null
+
+  submitWorkService =
+    work: {}
```
Comment (vietnogi) [inline src/scripts/services/submit-work.service.coffee]: This appears to be a global style implementation, which I strongly discourage.

File: src/scripts/services/submit-work.service.coffee [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+  submitWorkService.save = (updates) ->
+    console.log updates
+    apiCall = (model) ->
+      params =
+        id: currentWorkId
+
+      SubmitWorkAPIService.put(params, model).$promise
+
+    Optimist.update {
```
Comment (vietnogi) [inline src/scripts/services/submit-work.service.coffee]: don't think you need the brackets

File: src/scripts/services/submit-work.service.coffee [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    if workId != currentWorkId
+      submitWorkService.work = {}
+      currentWorkId = workId
+
+    params =
+      id: currentWorkId
+    apiCall = () ->
+      SubmitWorkAPIService.get(params).$promise
+
+    Optimist.fetchOne {
```
Comment (vietnogi) [inline src/scripts/services/submit-work.service.coffee]: don't need brackets

File: src/scripts/services/submit-work.service.coffee (hunk 5 above)
Comment (samsep) [inline src/scripts/services/submit-work.service.coffee]: might need to sync up with @aselbie on this and the $rootScope issue

**Pull request merged**